### PR TITLE
[Enhancement] - Choose Detection Method from Argv

### PR DESCRIPTION
**Pull request by FaaizHaikal**
## Description

Choose detection methods based on argv input.

## Type of Change

- [ ] Bugfix
- [x] Enhancement
- [ ] New feature
- [ ] Breaking change (fix or feature that would cause the existing functionality to not work as expected)

## How Has This Been Tested?

- [ ] New unit tests added.
- [x] Manual tested.

## Checklist:

- [x] Using Branch Name Convention
    - `feature/JIRA-ID-SHORT-DESCRIPTION` if has a JIRA ticket
    - `enhancement/SHORT-DESCRIPTION` if has/has no JIRA ticket and contain enhancement
    - `hotfix/SHORT-DESCRIPTION` if the change doesn't need to be tested (urgent)
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have made the documentation for the corresponding changes.

### CODE REVIEW

File: include/ninshiki_cpp/node/ninshiki_cpp_node.hpp [hunk 1]
```diff
@@ -81,7 +76,7 @@ class NinshikiCppNode
   ConfigGrpc config_grpc;
   std::string path;
 
-  static std::string get_node_prefix();
+  static const std::string get_node_prefix();
```
Comment (threeal) [inline include/ninshiki_cpp/node/ninshiki_cpp_node.hpp:79]: No need to specify `const` in here. Actualy, it would be better to pass it as a `std::stringview`. But nevermind, we shouldn't use this function, specifying the node prefix directly is better.

File: include/ninshiki_cpp/node/ninshiki_cpp_node.hpp [hunk 2]
```diff
@@ -49,15 +49,10 @@ class NinshikiCppNode
 
   NinshikiCppNode(
     rclcpp::Node::SharedPtr node, const std::string & path,
-    int frequency, shisen_cpp::Options options,
-    std::shared_ptr<DnnDetector> dnn_detection,
+    int frequency, std::shared_ptr<DnnDetector> dnn_detection,
```
Comment (threeal) [inline include/ninshiki_cpp/node/ninshiki_cpp_node.hpp:52]: It's actually not related with what you're trying to change, but it's better to pass an `std::shared_ptr` as a constant reference.

File: include/ninshiki_cpp/node/ninshiki_cpp_node.hpp [hunk 3]
```diff
@@ -21,6 +21,12 @@
 #ifndef NINSHIKI_CPP__NODE__NINSHIKI_CPP_NODE_HPP_
 #define NINSHIKI_CPP__NODE__NINSHIKI_CPP_NODE_HPP_
 
+#include <memory>
+#include <opencv2/dnn.hpp>
+#include <opencv2/highgui.hpp>
+#include <opencv2/imgproc.hpp>
+#include <string>
```
Comment (threeal) [inline include/ninshiki_cpp/node/ninshiki_cpp_node.hpp:28]: You don't need to change this, but please avoid changing stuffs that is not related with what you're trying to change (e.g. refactoring or formatting codes). It's better to create another pull request for that.

File: src/ninshiki_cpp/detector/dnn_detector.cpp [hunk 4]
```diff
@@ -38,15 +38,20 @@ DnnDetector::DnnDetector(bool gpu, bool myriad)
...
     classes.push_back(line);
   }
 
+
+}
+
+void DnnDetector::set_computation_method(bool gpu, bool myriad)
+{
+  this->gpu = gpu;
+  this->myriad = myriad;
```
Comment (threeal) [inline src/ninshiki_cpp/detector/dnn_detector.cpp:53]: Don't change this in here, but having an enum is better than using these variables.

File: src/ninshiki_cpp_detector.cpp [hunk 5]
```diff
@@ -18,36 +18,39 @@
...
   std::string topic_name = "";
-  std::string detection_method = "yolo";
+
   int gpu = 0;
   int myriad = 0;
   int frequency = 96;
 
+  std::shared_ptr<ninshiki_cpp::detector::DnnDetector>dnn_detector = nullptr;
+  std::shared_ptr<ninshiki_cpp::detector::ColorDetector>color_detector = nullptr;
+  std::shared_ptr<ninshiki_cpp::detector::LBPDetector>lbp_detector = nullptr;
```
Comment (threeal) [inline src/ninshiki_cpp_detector.cpp:41]: All of them are already derived from the `ninshiki_cpp::detector::Detector` class, it would be better to use that instead for all of them:
```suggestion
  std::shared_ptr<ninshiki_cpp::detector::Detector> detector = nullptr;
```

If you're stuck with handling the `DnnDetector::set_computation_method` or `ColorDetector::load_configuration` method, then simply just put that in the base class and make the implementation for the other class to do nothing.

File: src/ninshiki_cpp_detector.cpp [hunk 6]
```diff
@@ -87,35 +81,48 @@ int main(int argc, char ** argv)
...
           std::cout << "Unknown argument `" << arg << "`!\n\n" << help_message << std::endl;
           return 1;
         }
-      } else if (pos == 0) {
+      } else if (arg == "dnn") {
+        dnn_detector = std::make_shared<ninshiki_cpp::detector::DnnDetector>();
+      } else if (arg == "color") {
+        color_detector = std::make_shared<ninshiki_cpp::detector::ColorDetector>();
+      } else if (arg == "lbp") {
+        lbp_detector = std::make_shared<ninshiki_cpp::detector::LBPDetector>();
```
Comment (threeal) [inline src/ninshiki_cpp_detector.cpp:100]: Seriously you should utilize [argparse](https://github.com/p-ranav/argparse), but don't do that here, just do that on a separate PR when you have the time in the future.

**Review by threeal — CHANGES_REQUESTED** [review]

File: src/ninshiki_cpp_detector.cpp (hunk 5 above)
Comment (FaaizHaikal) [reply src/ninshiki_cpp_detector.cpp:41]: the program should allow more than one detection method. Should I use vector of ```std::shared_ptr<ninshiki_cpp::detector::Detector> detector``` with fixed size 3 instead?

Comment (threeal) [reply src/ninshiki_cpp_detector.cpp:41]: Aah i just realized what this program trying to do with the detectors, if it ended up going to be passed separately to the ninshiki_cpp::node::NinshikiCppNode` then just keep the code like this for now. Forget my previous suggestion.

**Review by threeal — APPROVED** [review]

**Review by hiikariri — APPROVED** [review]
LGTM
